### PR TITLE
GRPCD: A service daemon for a farcaster grpc api

### PR DESCRIPTION
**Pull request by TheCharlatan**
This completes the rough skeleton of a farcaster grpc service. I would like to get this in early, while the entire api is not complete yet to make subsequent review easier and to receive feedback while the code is still as small as possible. Once this is merged, I will open issues to track the implementation of the rest of the API, as well as better functional tests again the grpc server.

Note that the architecture implemented now can be easily re-used or replaced with another standard server, like a pure http server, or a json rpc server. 

~The protobuf generated file is still in a non-optimal location `src/farcaster.rs`. Not sure where it should go tbh.~
The protobuf generated file is now in the build directory.

### CODE REVIEW

**Comment by TheCharlatan** [comment]
@h4sh3d  and @zkao any idea what bumped the edition limitation that the CI is complaining about? It compiles find on my machine.

**Comment by h4sh3d** [comment]
> @h4sh3d and @zkao any idea what bumped the edition limitation that the CI is complaining about? It compiles find on my machine.

Checking dependencies need with
```
cargo msrv list
Fetching index
┌────────┬────────────────────────────────────────────────────────────────────────────────────┐
│ MSRV   ┆ Dependency                                                                         │
╞════════╪════════════════════════════════════════════════════════════════════════════════════╡
│        ┆ farcaster_node, zmq, tonic-build, tonic, toml, sysinfo, strict_encoding_derive,    │
│        ┆ strict_encoding, slip132, shellexpand, serde_yaml, serde_with, serde_json, serde,  │
│        ┆ rustc-hex, prost, paste, nix, monero-rpc, monero-lws, monero, microservices, log,  │
│        ┆ lnpbp, lnp-core, lightning_encoding, lazy_static, internet2, hex, farcaster_core,  │
│        ┆ env_logger, electrum-client, dotenv, configure_me_codegen, configure_me, config,   │
│        ┆ config, colored, clap_generate, clap, chrono, bitcoincore-rpc, bitcoin, base64,    │
│        ┆ anyhow, amplify_derive, amplify, zmq-sys, libc, bitflags, quote, prost-build,      │
│        ┆ tracing-futures, tracing, tower-service, tower-layer, tokio-stream, prost-derive,  │
│        ┆ percent-encoding, hyper-timeout, hyper, http-body, h2, bytes, base64, axum,        │
│        ┆ winapi, socket2, signal-hook-registry, pin-project-lite, parking_lot, once_cell,   │
│        ┆ num_cpus, mio, memchr, rayon, ntapi, doc-comment, core-foundation-sys, cfg-if,     │
│        ┆ encoding_derive_helpers, amplify_syn, miniscript, bitcoin_hashes, bitcoin,         │
│        ┆ dirs-next, yaml-rust, linked-hash-map, dtoa, serde_with_macros, rustversion, ryu,  │
│        ┆ itoa, serde_derive, cc, uuid, reqwest, jsonrpc-core, fixed-hash, tiny-keccak,      │
│        ┆ thiserror, serde-big-array, sealed, hex-literal, curve25519-dalek, base58-monero,  │
│        ┆ lnpbp_elgamal, lnpbp_chain, lnpbp_bech32, base64-compat, base58, secp256k1,        │
│        ┆ descriptor-wallet, lightning_encoding_derive, url, inet2_derive, inet2_addr,       │
│        ┆ chacha20poly1305, sha3, sha2, secp256kfun, rand_chacha, rand, rand, ecdsa_fun,     │
│        ┆ bitvec, bincode, termcolor, regex, humantime, atty, webpki-roots, webpki, socks,   │
│        ┆ rustls, void, unicode-segmentation, man, fmt2io, cargo_toml, parse_arg,            │
│        ┆ serde-hjson, rust-ini, nom, vec_map, textwrap, strsim, os_str_bytes, indexmap,     │
│        ┆ clap_derive, time, num-traits, num-integer, jsonrpc, bitcoincore-rpc-json, bech32, │
│        ┆ stringly_conversions, amplify_num, zeromq-src, metadeps, which, tempfile,          │
│        ┆ prost-types, petgraph, multimap, itertools, heck, cmake, tracing-core,             │
│        ┆ tracing-attributes, slab, pin-project-internal, tokio-io-timeout, want, httpdate,  │
│        ┆ httparse, fnv, pin-utils, sync_wrapper, mime, matchit, axum-core,                  │
│        ┆ winapi-x86_64-pc-windows-gnu, winapi-i686-pc-windows-gnu, parking_lot_core,        │
│        ┆ lock_api, hermit-abi, wasi, rayon-core, either, crossbeam-deque, autocfg,          │
│        ┆ bitcoin_hashes, bech32, dirs-sys-next, darling, getrandom, winreg, web-sys,        │
│        ┆ wasm-bindgen-futures, wasm-bindgen, tokio-native-tls, serde_urlencoded,            │
│        ┆ native-tls, js-sys, ipnet, hyper-tls, encoding_rs, static_assertions, byteorder,   │
│        ┆ crunchy, thiserror-impl, heck, zeroize, subtle, rand_core, digest, inflate,        │
│        ┆ deflate, secp256k1-sys, rand, slip132, psbt, commit_verify, bitcoin_scripts,       │
│        ┆ bitcoin_hd, matches, idna, form_urlencoded, torut, ed25519-dalek, poly1305,        │
│        ┆ cipher, chacha20, aead, opaque-debug, keccak, block-buffer, cpufeatures,           │
│        ┆ subtle-ng, secp256kfun_parity_backend, rand_core, ppv-lite86, rand_hc, rand_hc,    │
│        ┆ rand_chacha, getrandom, sigma_fun, wyz, tap, radium, funty, winapi-util,           │
│        ┆ regex-syntax, aho-corasick, quick-error, untrusted, ring, ws2_32-sys, winapi, sct, │
│        ┆ base64, roff, serde, num-traits, linked-hash-map, version_check, lexical-core,     │
│        ┆ unicode-width, hashbrown, proc-macro-error, serde_str_helpers, toml, pkg-config,   │
│        ┆ error-chain, remove_dir_all, redox_syscall, fixedbitset, try-lock,                 │
│        ┆ http-range-header, smallvec, scopeguard, windows_x86_64_msvc, windows_x86_64_gnu,  │
│        ┆ windows_i686_msvc, windows_i686_gnu, windows_aarch64_msvc, crossbeam-utils,        │
│        ┆ crossbeam-channel, crossbeam-epoch, redox_users, darling_macro, darling_core,      │
│        ┆ wasi, wasm-bindgen-macro, security-framework-sys, security-framework, schannel,    │
│        ┆ openssl-sys, openssl-probe, openssl, zeroize_derive, generic-array, adler32,       │
│        ┆ rand_xorshift, rand_pcg, rand_os, rand_jitter, rand_isaac, rand_hc, rand_core,     │
│        ┆ rand_chacha, autocfg, descriptors, unicode-normalization, unicode-bidi, hmac,      │
│        ┆ derive_more, base32, ed25519, universal-hash, cpuid-bool, block-padding, wasi,     │
│        ┆ curve25519-dalek-ng, spin, winapi-build, serde_test, arrayvec,                     │
│        ┆ proc-macro-error-attr, memoffset, ident_case, wasm-bindgen-macro-support,          │
│        ┆ core-foundation, vcpkg, foreign-types, synstructure, typenum, rand_core, rdrand,   │
│        ┆ fuchsia-cprng, cloudabi, tinyvec, crypto-mac, rustc_version, convert_case,         │
│        ┆ signature, wasm-bindgen-shared, wasm-bindgen-backend, foreign-types-shared,        │
│        ┆ unicode-xid, tinyvec_macros, semver, bumpalo, semver-parser, pest, ucd-trie        │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.31.0 ┆ syn, proc-macro2, unicode-ident                                                    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.36.0 ┆ futures-core, futures-sink, itoa, futures-task, futures-io                         │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.37.0 ┆ pin-project                                                                        │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.39.0 ┆ async-trait                                                                        │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.45.0 ┆ futures-util, async-stream, futures-channel, futures-macro, async-stream-impl,     │
│        ┆ futures-executor, futures                                                          │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.46.0 ┆ tokio-macros, windows-sys                                                          │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.49.0 ┆ tokio, tower, tokio-util, http                                                     │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.51.0 ┆ tower-http                                                                         │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1.56.0 ┆ prettyplease                                                                       │
└────────┴────────────────────────────────────────────────────────────────────────────────────┘
```
Shows that `prettyplease` cannot compile below `1.56.0` while our current MSRV is `1.55.0`.

This will be solved after https://github.com/farcaster-project/farcaster-node/pull/404 is merged and this is rebased as it bumps MSRV to `1.58.1`.

**Comment by TheCharlatan** [comment]
The functional tests fail after attempting to parse the node info from the cli. Since the tests pass fine locally, I am guessing that this is a caching issue.

**Comment by h4sh3d** [comment]
Tests pass, one post-CI step got canceled. I will restart it but we can consider CI ok.

File: src/config.rs [hunk 1]
```diff
@@ -96,11 +109,16 @@ impl Default for Config {
 pub struct FarcasterdConfig {
     /// Sets the auto-funding parameters, default to no auto-fund
     pub auto_funding: Option<AutoFundingConfig>,
+    /// Sets the grpc server port, if none is given, no grpc server is run
+    pub grpc_port: Option<u64>,
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/config.rs]: yeah, this implementation was lazy. Done in 411890dc29f4787b9a8cd1d87ed4202f5fb3ed61

File: src/farcasterd/runtime.rs [hunk 2]
```diff
@@ -400,6 +415,7 @@ impl Runtime {
     ) -> Result<(), Error> {
         match (&request, &source) {
             (Request::Hello, _) => {
+                debug!("Hello farcasterd from {}", source);
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/farcasterd/runtime.rs]: Done 8fafb4d253a55d1f2ae1dbd481da31cf9945453c

File: src/grpcd/opts.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+/// Walletd daemon; part of Farcaster Node
+#[derive(Parser, Clone, PartialEq, Eq, Debug)]
+#[clap(name = "grpcd", bin_name = "grpcd", author, version)]
+pub struct Opts {
+    /// These params can be read also from the configuration file, not just
+    /// command-line args or environment variables
+    #[clap(flatten)]
+    pub shared: crate::opts::Opts,
+
+    /// Trade role of participant (Maker or Taker)
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/grpcd/opts.rs]: Done 8fafb4d253a55d1f2ae1dbd481da31cf9945453c

File: src/grpcd/runtime.rs [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+        tx_response,
+    };
+
+    let mut service = Service::service(config, runtime)?;
+    service.add_loopback(rx_request)?;
+    service.run_loop()?;
+    unreachable!()
+}
+
+use farcaster::farcaster_server::{Farcaster, FarcasterServer};
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/grpcd/runtime.rs]: Done 8fafb4d253a55d1f2ae1dbd481da31cf9945453c

File: src/rpc/request.rs [hunk 5]
```diff
@@ -390,7 +390,7 @@ pub enum Request {
     // Can be issued from `cli` to `lnpd`
     #[api(type = 100)]
     #[display("get_info()")]
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/rpc/request.rs]: No Display derivable for Option, so instead I'll use display(Debug). Done in daa79ae08676fee2db739b3be4e34982da4acec5

**Comment by TheCharlatan** [comment]
> Is it necessary to commit generated file `src/farcaster.rs`?

No, I removed it now. If it gives us any trouble down the road we can always add it back.

File: src/bin/grpcd.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    opts.process();
+    trace!("Processed arguments: {:?}", &opts);
+
+    let service_config: ServiceConfig = opts.shared.clone().into();
+    trace!("Daemon configuration: {:#?}", &service_config);
+    debug!("MSG RPC socket {}", &service_config.msg_endpoint);
+    debug!("CTL RPC socket {}", &service_config.ctl_endpoint);
+
+    debug!("Starting runtime ...");
+    grpcd::run(service_config, opts.grpc_port).expect("Error running walletd runtime");
```
Comment (zkao) [inline src/bin/grpcd.rs:49]: s/wallet/grpc

**Review by zkao — APPROVED** [review]
looks good to me

File: src/bin/grpcd.rs (hunk 6 above)
Comment (TheCharlatan) [reply src/bin/grpcd.rs]: :((( Done in ca1757b8badb3f87ea3bd6b7f62f2d5039dd58a8

**Review by h4sh3d — APPROVED** [review]
This PR is ready to go. Nonetheless I open a series of issues to improve non-functional part of this work so we can debate on them separately:  #466,  #467, #468.

**Pull request merged**